bigtable/beam/workload-generator/src/main/java/bigtable/WorkloadGenerator.java

```diff
@@ -88,19 +81,47 @@ private static void cancelJob(BigtableWorkloadOptions options, DataflowPipelineJ
     client.updateJob(jobId, job);
   }
 
-  public static class ReadFromTableFn extends AbstractCloudBigtableTableDoFn<Long, Void> {
+  public static class ReadFromTableFn extends DoFn<Long, Void> {
+
+    static Object lock = new Object();
+    static BigtableDataClient bigtableDataClient;
+    static int refCount = 0;
+
+    private final String projectId;
+    private final String instanceId;
+
+    public ReadFromTableFn(String projectId, String instanceId) {
+      this.projectId = projectId;
+      this.instanceId = instanceId;
+    }
+
+    @Setup
+    void setup() throws IOException {
+      synchronized (lock) {
+        if (++refCount == 1) {
+          bigtableDataClient = BigtableDataClient.create(this.projectId, this.instanceId);
```

Comment:
also, you might want to add a catch to rollback the refCount if you failed to instantiate the client

Reply:
If the client fails to initiate won't the whole program exit from the exception? I could add the catch and just print that the connection failed, but for the purposes of this example I feel like it is fine to just fail?